net: lib: lwm2m: offer config parameter to reuse LWM2M registration

chmielewskiandreas opened the pull request:
To avoid a new registration on every connection drop and therefore data
traffic, try to use an existing LWM2M registration.

Hello reviewers,
think the idea is quite clear - if the connection drops, there is no reason to do a full LWM2M registration all the time. 

This patch is just an proposal of what I want to achieve. The changes works but has an open issue:

```
[00:00:15.239,471] <err> net_lwm2m_engine: Unable to resolve address
[00:00:15.239,471] <err> net_lwm2m_rd_client: Cannot init LWM2M engine (-2)
[00:00:15.239,471] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:15.239,501] <inf> net_lwm2m_rd_client: ***** ENGINE_NETWORK_ERROR
[00:00:15.739,624] <inf> net_lwm2m_rd_client: ------------> new state
[00:00:15.739,685] <inf> net_lwm2m_rd_client: ***** ENGINE_NETWORK_ERROR
ct> modem_ctrl generic_at AT+CFUN=1
Sending "AT+CFUN=1"
[00:00:18.539,154] <dbg> ct: handle_reg_status: Registration status change received: 2
[00:00:18.539,215] <dbg> ct: handle_cell_update: Old cell id: 0xffffffff 
[00:00:18.539,245] <wrn> ct: New cell id: 0x00ad1601 
[00:00:18.539,245] <dbg> ct: handle_cell_update: Old tracking area code: 0xffffffff 
[00:00:18.539,245] <wrn> ct: New tracking area code: 0xac0b 
[00:00:18.587,005] <dbg> ct: lte_event_cb: LTE_LC_EVT_RRC_UPDATE: 1
[00:00:19.240,539] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:19.240,600] <inf> net_lwm2m_rd_client: ***** ENGINE_DO_REGISTRATION
[00:00:19.740,722] <inf> net_lwm2m_rd_client: ------------> new state
[00:00:19.740,783] <inf> net_lwm2m_rd_client: ***** ENGINE_DO_REGISTRATION
[00:00:19.740,905] <inf> net_lwm2m_rd_client: RD Client started with endpoint '89882390000071092269' with client lifetime 60
[00:00:19.741,271] <err> net_lwm2m_engine: Unable to resolve address
[00:00:19.741,302] <err> net_lwm2m_rd_client: Cannot init LWM2M engine (-2)
[00:00:19.741,302] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:19.741,302] <inf> net_lwm2m_rd_client: ***** ENGINE_NETWORK_ERROR
[00:00:19.828,491] <dbg> ct: handle_reg_status: Registration status change received: 5
[00:00:19.828,857] <wrn> ct: AT notification received: +CNEC_ESM: 50,0
[00:00:20.240,600] <inf> net_lwm2m_rd_client: ------------> new state
[00:00:20.240,631] <inf> net_lwm2m_rd_client: ***** ENGINE_NETWORK_ERROR
[00:00:27.742,492] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:27.742,553] <inf> net_lwm2m_rd_client: ***** ENGINE_DO_REGISTRATION
[00:00:28.242,675] <inf> net_lwm2m_rd_client: ------------> new state
[00:00:28.242,706] <inf> net_lwm2m_rd_client: ***** ENGINE_DO_REGISTRATION
[00:00:28.242,858] <inf> net_lwm2m_rd_client: RD Client started with endpoint '89882390000071092269' with client lifetime 60
[00:00:28.258,819] <inf> net_lwm2m_rd_client: *************** reusing registration
[00:00:28.258,850] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:28.258,850] <inf> net_lwm2m_rd_client: ***** ENGINE_REGISTRATION_DONE
[00:00:28.258,850] <dbg> ct: lwm2m_event_cb: Registration complete
[00:00:28.759,094] <inf> net_lwm2m_rd_client: ------------> new state
[00:00:28.759,124] <inf> net_lwm2m_rd_client: ***** ENGINE_REGISTRATION_DONE
[00:00:28.759,490] <inf> net_lwm2m_rd_client: ************** set_sm_state
[00:00:28.759,490] <inf> net_lwm2m_rd_client: ***** ENGINE_UPDATE_SENT
[00:00:29.023,620] <inf> net_lwm2m_rd_client: Update callback (code:2.4)
```
You see that once the connection drops, the state machine changes to state ENGINE_DO_REGISTRATION where a new registration is tried. If the connection comes back an event 'ENGINE_REGISTRATION_DONE' is sent. The state changes are a bit misleading. A proposal is: In case of an successful registration, a connection drop should result in a new state where most of the code of the function 'sm_do_registration' is executed but without actually doing a new registration.

I am open for discussion :)

andreas-chmielewski-gcx commented:
Another issue: If the connection drops, the RD client is in a infinite loop trying to establish a connection. Is this on purpose and the user has to monitor the RD client and stop it at given time? The current behavior is quite bad for battery powered devices as we can not insist on having always a stable/good connection and if I understand the problem correctly this behavior also does not give the device time to get into PSM state to save battery power (the cpu should be idle for this).

bbilas commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -205,6 +205,16 @@ config LWM2M_RD_CLIENT_SUPPORT
 
+config LWM2M_RD_CLIENT_REUSE_REGISTRATION
+	bool "Reuse an existing LWM2M registration in case of an abort to LWM2M server"
+	default n
```
That is not needed, please remove it. Please see https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#redundant-defaults

chmielewskiandreas commented:
> Another issue: If the connection drops, the RD client is in a infinite loop trying to establish a connection. Is this on purpose and the user has to monitor the RD client and stop it at given time? The current behavior is quite bad for battery powered devices as we can not insist on having always a stable/good connection and if I understand the problem correctly this behavior also does not give the device time to get into PSM state to save battery power (the cpu should be idle for this).

I see that there is a retry mechanism ... I will create a new PR for that

rlubos commented:
CC @jheiskan81

andreas-chmielewski-gcx committed:
net: lib: lwm2m: RDClient: move update registration to a separate state

This patch is required to be able separate full registration from
updating registration when starting the RDClient.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

andreas-chmielewski-gcx committed:
net: lib: lwm2m: do a full registration of RDClient in a separate state

This patch is required to be able separate full registration from
updating registration when starting the RDClient.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

andreas-chmielewski-gcx committed:
net: lib: lwm2m: RDClient: try to reuse registration

The RDCLient performs currently everytime a full registration if e.g.
a network issues occurs. This is usually not necessary and should be
in particular when using battery powered NB-IoT devices. Updating the
registration should be sufficient in most of the cases if the lifetime
of the registration did not yet expired.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

andreas-chmielewski-gcx commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -169,6 +173,9 @@ static void set_sm_state(uint8_t sm_state)
 		event = LWM2M_RD_CLIENT_EVENT_DISCONNECT;
 	} else if (sm_state == ENGINE_NETWORK_ERROR) {
+#if defined(CONFIG_LWM2M_RD_CLIENT_REUSE_REGISTRATION)
+		client.reuse_registration = false;
```
You are absolutely right. I have changed the code completely by inserting proper states into the state machine of the RDClient. Thx for refering to a previous PR. I am actually having some of the issues in my backlog that were addressed in that PR.

chmielewskiandreas replied:
You are absolutely right. I have changed the code completely by inserting proper states into the state machine of the RDClient. Thx for refering to a previous PR. I am actually having some of the issues in my backlog that were addressed in that PR.

chmielewskiandreas commented:
Hello, I want to ask for updates on this PR.

rlubos reviewed (commented):
Sorry for the delay, I've left a comment but apparently did not submit the review :(

rlubos commented:
Also, apparently the code doesn't build right now, please check the twister results.

chmielewskiandreas committed:
Merge branch 'main' into reuse_registration

SeppoTakalo commented:
> I agree that observations must be kept. I will have to check what is needed to fix this but as far as I see, this needs deeper changes in how lwm2m engine is started.

I think you are starting to realize the point that I'm concerned.
I'm not happy to see the changes that try to send update packets when the engine is not 100% sure that update should work, and we have all the knowledge stored from previous session. The state machine of LwM2M engine is not very clear and not documented, so we might easily introduce new side effects if the change the flow. How did we end up in the de-registered state? If  we end up there, after succesfull deregistration, then definitely we cannot update, but we should register.

For really optimizing the registration and update, there is lots of work ongoing on queue-mode and sleeping. This should be the correct path. When we wake up from the sleep in the queue mode, we start with the registration update, and only if it fails, we back up into the registration.

If the aim is to improve the handling of the network error, so that we don't back off into de-register state so easily, then those should be done into the error handling. Not in the registration handling.

huelsenfruchtzwerg commented:
Butting in, since I've been in close exchange with @chmielewskiandreas regarding this feature.

>  Because it looks like you assume that your identity stays, even when you have somehow dropped the connection. How about if IP changes? How about if server have tried to connect and have dropped the connection.

IP could of course change after a dropped connection, but AFAICT it's intended by LwM2M to be able to update the IP with a registration update. Same with DTLS, as long as the server validates the DTLS-Identitity (which it must) there's nothing preventing us from updating the registration with a new DTLS-Session (though DTLS session resumption would of course be preferred, but that's a separate issue).

>  I'm not happy to see the changes that try to send update packets when the engine is not 100% sure that update should work, and we have all the knowledge stored from previous session. The state machine of LwM2M engine is not very clear and not documented, so we might easily introduce new side effects if the change the flow. How did we end up in the de-registered state? If we end up there, after succesfull deregistration, then definitely we cannot update, but we should register.

We should try to be smarter about when to register vs trying to send a registration update, but the concept is sound and imho supported by a reading of the lwm2m standard. There's nothing in the standard that specifies that a registration should be discarded on network errors/connection loss, with or without queue mode. To the contrary, the standard even specifies that it SHOULD be retained even across reboots:

`In order to avoid network traffic increase, the LwM2M Client SHOULD maintain previous observation states in case of
reboot, power cycle.`

IMHO the rd_client should use the last_updated + registration lifetime variables to figure out, whether the registration might still be valid and decide whether to update or register. A de-registration or if the server denies the registration update could just clear those variables to indicate that no valid registration is active. In all other cases, (including timeouts, network errors, etc) as long as the lifetime of the old registration is not expired, we SHOULD strife to update it instead of needlessly re-registering.

> For really optimizing the registration and update, there is lots of work ongoing on queue-mode and sleeping. This should be the correct path. When we wake up from the sleep in the queue mode, we start with the registration update, and only if it fails, we back up into the registration.

You're correct, that this PR and a proper Queue-Mode implementation share a lot of concerns. Is there already a PR/branch with a proposed implementation somewhere we could have a look at?
If not so, I don't see why we shouldn't implement a "proper" handling of the registration state now already (not saying that that is this PR in it's current version, but we can discuss how a proper approach would look here and then implement what we agree on).

This requires some care of course, but overall shouldn't be that hard imho. 
I think mostly it's just necessary to move all clearing of registration state from fault_cbs/timeout_cbs etc to the point where the new registration is performed. Or do you disagree?

SeppoTakalo commented:
Let me try to clarify this a bit.

I'm not against of doing UPDATE to renew your state on the server in case you have had your connection problems.

I think this PR tries to solve the problem by modifying that wrong path of the state machine:
https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events
![image](https://user-images.githubusercontent.com/3104794/173319901-4593146d-5433-464c-92dd-08363b87b3c7.png)

Your intention was to modify the event where we start the registration phase. But as you can see, we might have gotten there by proper de-registration, or just that it might be the initial state. So all those are correct paths. Don't alter those by introducing intermediate states.

But as you can see, from network errors, we also drop into same "DO REGISTER" state. There is a room for improvement there. If you can detect a cases like IP have changed, NAT mapping have changed or similar, then those have to be done there, before we drop into "DO REGISTER". Because if we drop into that state, we already have code there that drop all the server registered observations. If those are lost, then we should do the full registration. Even if the UPDATE message would go through, we are out of synch now, and that is bad.

Does this clarify it?
It is not the wrong outcome that this PR is after. It definitely is a correct move. But I think it is not implemented on right path of the state machine.

chmielewskiandreas commented:
> Let me try to clarify this a bit.
> 
> I'm not against of doing UPDATE to renew your state on the server in case you have had your connection problems.
> 
> I think this PR tries to solve the problem by modifying that wrong path of the state machine: https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events ![image](https://user-images.githubusercontent.com/3104794/173319901-4593146d-5433-464c-92dd-08363b87b3c7.png)
> 
> Your intention was to modify the event where we start the registration phase. But as you can see, we might have gotten there by proper de-registration, or just that it might be the initial state. So all those are correct paths. Don't alter those by introducing intermediate states.
> 
> But as you can see, from network errors, we also drop into same "DO REGISTER" state. There is a room for improvement there. If you can detect a cases like IP have changed, NAT mapping have changed or similar, then those have to be done there, before we drop into "DO REGISTER". Because if we drop into that state, we already have code there that drop all the server registered observations. If those are lost, then we should do the full registration. Even if the UPDATE message would go through, we are out of synch now, and that is bad.
> 
> Does this clarify it? It is not the wrong outcome that this PR is after. It definitely is a correct move. But I think it is not implemented on right path of the state machine.

@SeppoTakalo : Thanks for your comment and ideas. I really appreciate it! I have created a new PR that covers one corner case
https://github.com/zephyrproject-rtos/zephyr/pull/46687

chmielewskiandreas commented:
Any update/comments on my last commit?
